OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+  /// \param first Beginning of range covered by the iterator.
+  /// \param last End of the range covered by the iterator.
+  /// \param ncol Number of columns to create in the DataTable_. This is only
+  ///             used(and required) when the function is called on an empty 
+  ///             DataTable_. Ignore otherwise.
+  /// \param allow_missing Allow for missing values. When set to false, this
+  ///                      function will throw if the input iterator fills up 
+  ///                      the last row only partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws OpenSim::ZeroElements If the number of elements produced by the
+  ///                               input iterator is zero.
+  /// \throws OpenSim::InvalidEntry If the function is called on an empty
+  ///                               DataTable_ without providing the argument
+  ///                               ncol.
+  /// \throws OpenSim::InvalidEntry If allow_missing is false and the input
+  ///                               iterator does not fill up the last row
+  ///                               completely.
+  template<typename InputIt>
+  void addRows(InputIt first, 
+               InputIt last, 
+               const size_t ncol = std::numeric_limits<size_t>::max(),
+               const bool allow_missing = false) {
+    if(!(first != last))
+      throw ZeroElements{"Input iterators produce zero elements."};
+    if((data.nrow() == 0 || data.ncol() == 0) && 
+       (ncol == std::numeric_limits<size_t>::max() || ncol == 0))
+      throw InvalidEntry{"DataTable is empty. 'ncol' argument must be" 
+                         " provided and it cannot be zero."};
+
+    data.resizeKeep(data.nrow() + 1, 
+                    data.ncol() == 0 ? static_cast<int>(ncol) : data.ncol());
+    int row{data.nrow() - 1};
+    int col{0};
+    while(first != last) {
+      data.set(row, col, *first);
+      ++first; ++col;
+      if(col == static_cast<int>(data.ncol()) && first != last) {
+        col = 0;
+        ++row;
+        data.resizeKeep(data.nrow() + 1, data.ncol());
+      }
+    }
+    if(!allow_missing && col != data.ncol())
+      throw InvalidEntry{"Input iterator did not produce enough elements to "
+                         "fill the last row."};
+  }
+
+  /// Add(append) a col to the DataTable_ using a SimTK::Vector_. If the 
+  /// DataTable is empty, input col will be the first col. This function can be
+  /// used to populate an empty DataTable_.
+  ///
+  /// \param col The col to be added as a SimTK::Vector_.
+  ///
+  /// \throws OpenSim::ZeroElements If number of elements in the input col is
+  ///                               zero.
+  /// \throws OpenSim::InvalidEntry If number of columns in the input col does
+  ///                               not match the number of rows of the 
+  ///                               DataTable_.
+  void addCol(const SimTK::Vector_<ET>& col) {
```

Comment:
I would prefer to see column instead of col in method names throughout the interface. It is somewhat verbose, but it is much more readable.